pkg/scheduler/queue/tree_queueing_algorithms.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+// QueuingAlgorithm represents the set of operations specific to different approaches to queuing/dequeuing. It is
+// applied at the layer-level -- every Node at the same depth in a MultiQueuingAlgorithmTreeQueue shares the same QueuingAlgorithm,
+// including any state in structs that implement QueuingAlgorithm.
+type QueuingAlgorithm interface {
+	// addChildNode creates a child Node of parent. QueuePaths passed to enqueue functions are allowed to contain
+	// node names for nodes which do not yet exist; in those cases, we create the missing nodes. Implementers are
+	// responsible for adding the newly-created node to whatever state(s) keep track of nodes in the subtree, as well
+	// as updating any positional pointers that have been shifted by the addition of a new node.
```

Comment:
loose suggestion, mostly about the first part - just want to provide some more information about what methods call this and when/why

`// addChildNode creates a child Node of parent` is not quite true, we aren't creating any nodes here, just adding or updating internal state in response to a node being added.
I might even lean towards `somethingUpdateState` to mirror `dequeueUpdateState`


```suggestion
	// addChildNode is called by Tree implementations during EnqueueBackByPath and EnqueueFrontByPath,
	// if enqueuing to a QueuePath requires creating a Node which previously did not exist in the Tree.
	// addChildNode is called so that the QueuingAlgorithm may update its internal state to reflect the Node creation
```